src/game/client/components/binds.cpp

```diff
@@ -25,82 +25,133 @@ const char CBinds::s_aaDefaultBindValues[][32] = {
+{
+	switch(Modifier)
+	{
+		case MODIFIER_SHIFT:
+			return Key == KEY_LSHIFT || Key == KEY_RSHIFT;
+		case MODIFIER_CTRL:
+			return Key == KEY_LCTRL || Key == KEY_RCTRL;
+		case MODIFIER_ALT:
+			return Key == KEY_LALT;
+		case MODIFIER_NONE:
+		default:
+			return false;
+	}
+}
+
+bool CBinds::CBindsSpecial::OnInput(IInput::CEvent Event)
+{
+	int Mask = getMask(Input());
+	bool rtn = false;
+
+	// don't handle anything but FX and composed binds
+	if(Mask == 1 && !(Event.m_Key >= KEY_F1 && Event.m_Key <= KEY_F12) && !(Event.m_Key >= KEY_F13 && Event.m_Key <= KEY_F24))
+		return false;
+
+	for(int m = 0; m < MODIFIER_COUNT; m++)
+	{
+		if((Mask&(1 << m)) && m_pBinds->m_aaaKeyBindings[Event.m_Key][m][0] != 0)
+		{
+			int Stroke = 0;
+			if(Event.m_Flags&IInput::FLAG_PRESS)
+				Stroke = 1;
+
+			m_pBinds->GetConsole()->ExecuteLineStroked(Stroke, m_pBinds->m_aaaKeyBindings[Event.m_Key][m]);
+			rtn = true;
+		}
+	}
+	return rtn;
+}
 
-bool CBinds::OnInput(IInput::CEvent e)
+bool CBinds::OnInput(IInput::CEvent Event)
 {
-	// don't handle invalid events and keys that arn't set to anything
-	if(e.m_Key <= 0 || e.m_Key >= KEY_LAST || m_aaKeyBindings[e.m_Key][0] == 0)
+	int Mask = getMask(Input());
+
+	// don't handle invalid events and keys that aren't set to anything
+	if(Event.m_Key <= 0 || Event.m_Key >= KEY_LAST)
 		return false;
 
-	if(e.m_Flags&IInput::FLAG_PRESS)
-		Console()->ExecuteLineStroked(1, m_aaKeyBindings[e.m_Key]);
-	if(e.m_Flags&IInput::FLAG_RELEASE)
-		Console()->ExecuteLineStroked(0, m_aaKeyBindings[e.m_Key]);
-	return true;
+	bool rtn = false;
+	for(int m = 0; m < MODIFIER_COUNT; m++)
+	{
+		if((Mask&(1 << m)) && m_aaaKeyBindings[Event.m_Key][m][0] != 0)
+		{
+			if(Event.m_Flags&IInput::FLAG_PRESS)
+				Console()->ExecuteLineStroked(1, m_aaaKeyBindings[Event.m_Key][m]);
+			if(Event.m_Flags&IInput::FLAG_RELEASE)
+				Console()->ExecuteLineStroked(0, m_aaaKeyBindings[Event.m_Key][m]);
+			rtn = true;
+		}
+	}
+	return rtn;
 }
 
 void CBinds::UnbindAll()
 {
 	for(int i = 0; i < KEY_LAST; i++)
-		m_aaKeyBindings[i][0] = 0;
+		for(int m = 0; m < MODIFIER_COUNT; m++)
+			m_aaaKeyBindings[i][m][0] = 0;
 }
 
-const char *CBinds::Get(int KeyID)
+const char *CBinds::Get(int KeyID, int Modifier)
 {
 	if(KeyID > 0 && KeyID < KEY_LAST)
-		return m_aaKeyBindings[KeyID];
+		return m_aaaKeyBindings[KeyID][Modifier];
 	return "";
 }
 
-const char *CBinds::GetKey(const char *pBindStr)
+void CBinds::GetKey(const char *pBindStr, char aBuf[64])
 {
-	for(int KeyId = 0; KeyId < KEY_LAST; KeyId++)
+	aBuf[0] = 0;
+	for(int KeyID = 0; KeyID < KEY_LAST; KeyID++)
 	{
-		const char *pBind = Get(KeyId);
-		if(!pBind[0])
-			continue;
+		for(int m = 0; m < MODIFIER_COUNT; m++)
+		{
+			const char *pBind = Get(KeyID, m);
+			if(!pBind[0])
+				continue;
 
-		if(str_comp(pBind, pBindStr) == 0)
-			return Input()->KeyName(KeyId);
+			if(str_comp(pBind, pBindStr) == 0)
+				str_format(aBuf, 64, "key %s%s not found", Input()->KeyName(KeyID), GetModifierName(m));
```

Comment:
should use sizeof(aBuf)

Reply:
I get a
`src/game/client/components/binds.cpp:152:33: warning: ‘sizeof’ on array function parameter ‘aBuf’ will return size of ‘char*’ [-Wsizeof-array-argument]
     str_format(aBuf, sizeof(aBuf), "key %s%s not found", Input()->KeyName(KeyID), GetModifierName(m));`

when I do this. It does not seem like this would work.

Reply:
Ah right. Could you pass a pointer to the array and the size of the array as function parameters instead?

Reply:
@oy Done